doc/man/pilot-install-todo.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "pilot-install-todo" 1 "Copyright 1996-2007 FSF" "0.12.5" "PILOT-LINK"
.SH NAME
pilot-install-todo \- Updates the Palm ToDo list with one new entry.
.SH "SECTION"

.PP
pilot\-link: Userland conduits

.SH "SYNOPSIS"

.PP
 \fIpilot\-install\-todo\fR [\fB\-p\fR|\fB\-\-port\fR <\fBport\fR>] [\fB\-q\fR|\fB\-\-quiet\fR] [\fB\-\-version\fR] [\fB\-?\fR|\fB\-\-help\fR] [\fB\-\-usage\fR] [\fB\-P\fR|\fB\-\-priority\fR [\fB1|2|3|4|5\fR]] [\fB\-d\fR|\fB\-\-due\fR  \fBmm/dd/yyyy\fR] [\fB\-c\fR|\fB\-\-completed\fR] [\fB\-D\fR|\fB\-\-description\fR [\fBtitle\fR]] [\fB\-n\fR|\fB\-\-note\fR  \fBnote\fR] [\fB\-f\fR|\fB\-\-file\fR  \fIfilename\fR]

.SH "DESCRIPTION"

.PP
 \fIpilot\-install\-todo\fR allows the user to install one new ToDo list entry with specified paramters passed to \fIpilot\-install\-todo\fR and add a Note entry using text found in a filename <\fIfilename\fR> passed to \fIpilot\-install\-todo\fR\&.

.SH "OPTIONS"

.SS "pilot-install-todo options"

                        \fB\-P\fR,
                        \fB\-\-priority\fR \fBpriority\fR
                    
.PP
Specify the priority of the ToDo entry \- 1|2|3|4|5 (defaults to a priority of 1)\&.

.PP
 \fINOTE:\fR this value has changed from the default of 4 to the default of 1, just as a new ToDo entry on the Palm would set it\&.

                        \fB\-d\fR,
                        \fB\-\-due\fR \fBduedate\fR
                    
.PP
The due Date MM/DD/YYYY (defaults to blank, no due date if this is not specified)\&.

                        \fB\-c\fR, \fB\-\-completed\fR
                    
.PP
Mark the item complete (the default is to mark the item as incomplete, as the Palm would do it)\&.

                        \fB\-D\fR,
                        \fB\-\-description\fR \fBtitle\fR
                    
.PP
The title of the ToDo entry

                        \fB\-n\fR, \fB\-\-note\fR
                        \fBnote\fR
                    
.PP
The Note text you wish to add to this ToDo entry, if any\&.

                        \fB\-f\fR,
                        \fB\-\-file\fR \fIfilename\fR
                    
.PP
A local filename containing the Note text\&.

.SS "Conduit Options"

                        \fB\-p\fR, \fB\-\-port\fR
                        <\fBport\fR>
                    
.PP
Use device file <\fIport\fR> to communicate with the Palm handheld\&. If this is not specified, \fIpilot\-install\-todo\fR will look for the \fI $PILOTPORT \fR environment variable\&. If neither are found or supplied, \fIpilot\-install\-todo\fR will print the usage information\&.

                        \fB\-q\fR, 
                        \fB\-\-quiet\fR
                    
.PP
Suppress 'Hit HotSync button' message

                        \fB\-v\fR, \fB\-\-version\fR
                    
.PP
Display version of \fIpilot\-install\-todo\fR and exit without connecting\&.

.SS "Help Options"

                        \fB\-h\fR, \fB\-\-help\fR
                    
.PP
Display the help synopsis for \fIpilot\-install\-todo\fR and exit without connecting\&.

                        \fB\-\-usage\fR 
                    
.PP
Display a brief usage message and exit without connecting\&.

.SH "USAGE"

.PP
The program will connect to a target Palm handheld, and creates the ToDo entry to the Palm ToDo application, using the specified parameters and input from the supplied file passed as arguments on the commandline\&.

.SH "EXAMPLE"

.PP
 \fIpilot\-install\-todo\fR  \fB\-p\fR  \fI/dev/pilot\fR  \fB\-n\fR  \fB'Buy Milk'\fR  \fB\-D\fR  \fB'Go shopping, see note for items'\fR 

.SH "AUTHOR"

.PP
 \fIpilot\-install\-todo\fR was originally written by Martin Fick, based on code in \fIpilot\-install\-todos\fR by Robert A\&. Kaplan\&.

.PP
This manual page was converted to XML by Neil Williams <linux@codehelp\&.co\&.uk> 

.SH "REPORTING BUGS"

.PP
We have an online bug tracker\&. Using this is the only way to ensure that your bugs are recorded and that we can track them until they are resolved or closed\&. Reporting bugs via email, while easy, is not very useful in terms of accountability\&. Please point your browser to http://bugs\&.pilot\-link\&.org: \fIhttp://bugs.pilot-link.org\fR and report your bugs and issues there\&.

.SH "COPYRIGHT"

.PP
This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version\&.

.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&. See the GNU General Public License for more details\&.

.PP
You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc\&., 51 Franklin St, Fifth Floor, Boston, MA 02110\-1301, USA\&.

.SH "SEE ALSO"

.PP
 \fIpilot\-install\-todos\fR(1), \fIpilot\-link\fR(7)\&.



src/pilot-install-todo.c
/*
 * $Id: pilot-install-todo.c,v 1.27 2006/10/12 14:21:20 desrod Exp $ 
 *
 * install-todo.c:  Palm ToDo list installer
 *
 * Copyright 2002, Martin Fick, based on code in install-todos by Robert A.
 * 		   Kaplan
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

#include "pi-userland.h"
#include <stdio.h>
#include <stdlib.h>

#include "pi-source.h"
#include "pi-dlp.h"
#include "pi-todo.h"
#include "pi-header.h"

poptContext po;

#ifndef HAVE_BASENAME
/* Yes, this isn't efficient... scanning the string twice */
#define basename(s) (strrchr((s), '/') == NULL ? (s) : strrchr((s), '/') + 1)
#endif

#define DATE_STR_MAX 1000

char *strptime(const char *s, const char *format, struct tm *tm);


/***********************************************************************
 *
 * Function:    read_file
 *
 * Summary:     Reads a file from disk to import into ToDoDB.pdb
 *
 * Parameters:  Pointer to a filename
 *
 * Returns:     Text inside the file, of course
 *
 ***********************************************************************/
int read_file(char *filename, char **text)
{
	FILE	*f;
	size_t	filelen;

	f = fopen(filename, "r");
	if (f == NULL) {
		fprintf(stderr,"   ERROR: Could not open '%s': %s\n",
			filename,strerror(errno));
		return -1;
	}

	fseek(f, 0, SEEK_END);
	filelen = ftell(f);
	fseek(f, 0, SEEK_SET);

	*text = (char *) malloc(filelen + 1);
	if (*text == NULL) {
		fprintf(stderr,"   ERROR: Could not allocate memory.\n");
		return -1;
	}

	fread(*text, filelen, 1, f);

	return 0;
}


/***********************************************************************
 *
 * Function:    install_ToDo
 *
 * Summary:     Adds the entry to ToDoDB.pdb
 *
 * Parameters:
 *
 * Returns:     Nothing
 *
 ***********************************************************************/
void install_ToDo(int sd, int db, struct ToDo todo)
{

	pi_buffer_t *ToDo_buf;
	char  duedate[DATE_STR_MAX];

	printf("Indefinite:  %i\n", todo.indefinite);
	/* Assume DATE_STR_MAX is at least 40 or so. */
	if (todo.indefinite == 0) {
		strftime(duedate, DATE_STR_MAX, "%a %b %d %H:%M:%S %Z %Y", &todo.due);
	} else {
		strncpy(duedate,"<none>",DATE_STR_MAX);
	}
	printf("Due Date:    %s\n", duedate);
	printf("Priority:    %i\n", todo.priority);
	printf("Complete:    %i\n", todo.complete);
	printf("Description: %s\n", todo.description);
	printf("Note:        %s\n", todo.note);

	ToDo_buf = pi_buffer_new(0);
	pack_ToDo(&todo, ToDo_buf, todo_v1);

	dlp_WriteRecord(sd, db, 0, 0, 0, ToDo_buf->data, ToDo_buf->used, 0);

	pi_buffer_free(ToDo_buf);
	return;
}

int main(int argc, const char *argv[])
{
	int 	sd	= -1,
		db,
		po_err  = -1;

	char
		*due		= NULL,
		*description	= NULL,
		*note		= NULL,
		*filename	= NULL;
	int priority = 1;
	int completed = 0;

	struct 	PilotUser User;
	struct 	ToDo todo;

        const struct poptOption options[] = {
		USERLAND_RESERVED_OPTIONS
                { "priority",    'P', POPT_ARG_INT, &priority, 0, "The Priority (default is 1)", "[1|2|3|4|5]"},
                { "due",         'd', POPT_ARG_STRING, &due, 0, "The due Date of the item (default is no date)", "mm/dd/yyyy"},
                { "completed",   'c', POPT_ARG_NONE, &completed, 0, "Mark the item complete (default is incomplete)"},
		{ "description", 'D', POPT_ARG_STRING, &description, 0, "The title of the ToDo entry", "[title]"},
                { "note",        'n', POPT_ARG_STRING, &note, 0, "The Note(tm) text (a single string)", "<note>"},
                { "file",        'f', POPT_ARG_STRING, &filename, 0, "A local filename containing the Note text", "[filename]"},
		POPT_TABLEEND
        };


        po = poptGetContext("install-todo", argc, (const char **) argv, options, 0);
	poptSetOtherOptionHelp(po,
		" [-p port] <todo-options>\n\n"
		"   Updates the Palm ToDo list with a single new entry\n\n"
		"   Example:\n"
		"       -p /dev/pilot -n 'Buy Milk' -D 'Go shopping, see note for items'\n\n");

	if (argc < 2) {
		poptPrintUsage(po,stderr,0);
		return 1;
	}

        while ((po_err = poptGetNextOpt(po)) >= 0) {
		fprintf(stderr,"   ERROR: Unhandled option %d.\n",po_err);
		return 1;
	}

	if (po_err < -1) {
		plu_badoption(po,po_err);
	}

	if ((priority < 1) || (priority > 5)) {
		priority=1;
		fprintf(stderr,"   WARNING: Changed illegal priority to 1.\n");
	}

	/*  Setup some todo defaults */
	todo.indefinite  = 1;
	todo.priority    = priority;
	todo.complete    = completed;
	todo.description = description;
	if (filename) {
		int c = read_file(filename,&todo.note);
		if (c<0) {
			return 1;
		}
	}
	else {
		todo.note = note;
	}

	if (due != NULL) {
		strptime(due, "%m/%d/%Y", &todo.due);
		todo.indefinite = 0;
		todo.due.tm_sec  = 0;
		todo.due.tm_min  = 0;
		todo.due.tm_hour = 0;
	}

	sd = plu_connect();

        if (sd < 0)
	    return 1;

	/* Open the ToDo database, store access handle in db */
	if (dlp_OpenDB(sd, 0, 0x80 | 0x40, "ToDoDB", &db) < 0) {
		fprintf(stderr,"   ERROR: Unable to open ToDo Database on Palm.\n");
		dlp_AddSyncLogEntry(sd, "Unable to open ToDo Database.\n");
		pi_close(sd);
		return 1;
	}

	install_ToDo(sd, db, todo);

	dlp_CloseDB(sd, db);

	/* Tell the user who it is, with a different PC id. */
	User.lastSyncPC 	= 0x00010000;
	User.lastSyncDate = User.successfulSyncDate = time(0);
	dlp_WriteUserInfo(sd, &User);

	dlp_AddSyncLogEntry(sd, "Wrote ToDo entry to Palm.\n\n"
		"Thank you for using pilot-link.\n");

	dlp_EndOfSync(sd, 0);
	pi_close(sd);

	return 0;
}

/* vi: set ts=8 sw=4 sts=4 noexpandtab: cin */
/* ex: set tabstop=4 expandtab: */
/* Local Variables: */
/* indent-tabs-mode: t */
/* c-basic-offset: 8 */
/* End: */

